File: spark-riak-connector-java/src/main/java/com/basho/spark/connector/japi/SparkContextJavaFunctions.java
package com.basho.spark.connector.japi;
import com.basho.riak.client.core.query.Location;
import com.basho.riak.client.core.query.Namespace;
import com.basho.riak.client.core.query.RiakObject;
import com.basho.spark.connector.japi.rdd.RiakJavaPairRDD;
import com.basho.spark.connector.japi.rdd.RiakJavaRDD;
import com.basho.spark.connector.rdd.RiakRDD;
import com.basho.spark.connector.rdd.RiakRDD$;
import com.basho.spark.connector.util.RiakObjectConversionUtil;
import org.apache.spark.SparkContext;

import scala.Function2;
import scala.Tuple2;
import scala.reflect.ClassTag;
import scala.runtime.AbstractFunction2;


import java.io.Serializable;

import static com.basho.spark.connector.util.JavaUtil.getClassTag;

public class SparkContextJavaFunctions {
    protected static class ConversionFunction<T> extends AbstractFunction2<Location, RiakObject, T> implements Serializable {
        private final ClassTag<T> classTag;

        public ConversionFunction(ClassTag<T> classTag) {
            this.classTag = classTag;
        }

        @Override
        public T apply(Location l, RiakObject r) {
            return RiakObjectConversionUtil.from(l, r, classTag);
        }

        public static <T> ConversionFunction create(ClassTag<T> classTag){
            return new ConversionFunction(classTag);
        }
    };

    public final SparkContext sparkContext;

    SparkContextJavaFunctions(SparkContext sparkContext) {
        this.sparkContext = sparkContext;
    }

    public <T> RiakJavaRDD<T> toJavaRDD(RiakRDD<T> rdd, Class<T> targetClass) {
        return new RiakJavaRDD<>(rdd, getClassTag(targetClass));
    }

    public <T> RiakJavaRDD<T> riakBucket(Namespace ns, Class<T> valueClass) {
        return this.riakBucket(ns.getBucketNameAsString(), ns.getBucketTypeAsString(), valueClass);
    }

    public <T> RiakJavaRDD<T> riakBucket(String bucketName, String bucketType, Class<T> valueClass) {
        final ClassTag<T> classTag = getClassTag(valueClass);
        final String bucketTypeStr = bucketType == null || bucketType.isEmpty() ? "default" : bucketType;
        final RiakRDD<T> rdd = RiakRDD$.MODULE$.apply(sparkContext, bucketTypeStr, bucketName,
                ConversionFunction.create(classTag), classTag);
        return new RiakJavaRDD(rdd, classTag);
    }

    public <K, V> RiakJavaPairRDD<K, V> riakBucket(Namespace ns, Function2<Location, RiakObject, Tuple2<K, V>> convert, Class<K> keyClass, Class<V> valueClass) {
        return this.riakBucket(ns.getBucketNameAsString(), convert, ns.getBucketNameAsString(), keyClass, valueClass);
    }

    public <K, V> RiakJavaPairRDD<K, V> riakBucket(String bucketName, Function2<Location, RiakObject, Tuple2<K, V>> convert, String bucketType, Class<K> keyClass, Class<V> valueClass) {
        final ClassTag<K> kClassTag = getClassTag(keyClass);
        final ClassTag<V> vClassTag = getClassTag(valueClass);
        final String bucketTypeStr = bucketType == null || bucketType.isEmpty() ? "default" : bucketType;
        final RiakRDD<Tuple2<K, V>> rdd = RiakRDD$.MODULE$.apply(sparkContext, bucketTypeStr, bucketName, convert, kClassTag, vClassTag);
        return new RiakJavaPairRDD(rdd, kClassTag, kClassTag);
    }
}

File: connector/src/main/scala/com/basho/riak/spark/japi/SparkContextJavaFunctions.scala
/**
  * Copyright (c) 2015 Basho Technologies, Inc.
  *
  * This file is provided to you under the Apache License,
  * Version 2.0 (the "License"); you may not use this file
  * except in compliance with the License.  You may obtain
  * a copy of the License at
  *
  *   http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing,
  * software distributed under the License is distributed on an
  * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
  * KIND, either express or implied.  See the License for the
  * specific language governing permissions and limitations
  * under the License.
  */
package com.basho.riak.spark.japi

import java.io.Serializable
import com.basho.riak.client.core.query.{Namespace, RiakObject, Location}
import com.basho.riak.spark.japi.rdd.{RiakTSJavaRDD, RiakJavaPairRDD, RiakJavaRDD}
import com.basho.riak.spark.rdd._
import com.basho.riak.spark.rdd.connector.RiakConnector
import com.basho.riak.spark.rdd.mapper.ReadDataMapperFactory.DefaultReadDataMapper
import com.basho.riak.spark.rdd.mapper.{ReadPairValueDataMapper, ReadDataMapperFactory, ReadValueDataMapper}
import com.basho.riak.spark.util.JavaApiHelper._
import com.basho.riak.spark.util.RiakObjectConversionUtil
import org.apache.spark.SparkContext
import org.apache.spark.sql.types.StructType

import scala.reflect.ClassTag
import scala.runtime.AbstractFunction2

class SparkContextJavaFunctions(@transient val sc: SparkContext) extends Serializable {

  class ConversionFunction[V](implicit classTag: ClassTag[V])
    extends AbstractFunction2[Location, RiakObject, V] with Serializable {

    def apply(l: Location, r: RiakObject): V = RiakObjectConversionUtil.from(l, r)(classTag)
  }

  def toJavaRDD[T](rdd: RiakRDD[T], targetClass: Class[T]): RiakJavaRDD[T] = new RiakJavaRDD[T](rdd, getClassTag(targetClass))
  def toJavaPairRDD[K, V](rdd: RiakRDD[(K, V)], keyClass: Class[K], valueClass: Class[V]): RiakJavaPairRDD[K, V] =
    new RiakJavaPairRDD[K, V](rdd, getClassTag(keyClass), getClassTag(valueClass))

  def riakBucket[T](bucketName: String, bucketType: String = BucketDef.DefaultBucketType, rdmf: ReadDataMapperFactory[T]): RiakJavaRDD[T] = {
    val rdd = RiakRDD[T](sc, bucketType, bucketName, None, ReadConf(sc.getConf))(getClassTag(rdmf.targetClass), rdmf)
    toJavaRDD(rdd, rdmf.targetClass)
  }

  def riakBucket[T](bucketName: String, bucketType: String, valueClass: Class[T]): RiakJavaRDD[T] = {
    val rdmf = ReadValueDataMapper.factory(getClassTag(valueClass))
    riakBucket(bucketName, bucketType, rdmf)
  }

  def riakBucket[T](ns: Namespace, rdmf: ReadDataMapperFactory[T]): RiakJavaRDD[T] = riakBucket(ns.getBucketNameAsString, ns.getBucketTypeAsString, rdmf)


  def riakBucket[T](ns: Namespace, valueClass: Class[T]): RiakJavaRDD[T] = riakBucket(ns.getBucketNameAsString, ns.getBucketTypeAsString, valueClass)


  def riakBucket[Tuple2[String, AnyRef]](ns: Namespace): RiakJavaRDD[(String, Any)] =
    riakBucket(ns.getBucketNameAsString, ns.getBucketTypeAsString)

  def riakBucket[Tuple2[String, AnyRef]](bucketName: String, bucketType: String): RiakJavaRDD[(String, Any)] =
    riakBucket(bucketName, bucketType, DefaultReadDataMapper)

  def riakBucket[K <: String, V](bucketName: String, bucketType: String,
                                 keyClass: Class[K], valueClass: Class[V], rdmf: ReadDataMapperFactory[(K, V)]): RiakJavaPairRDD[K, V] = {
    toJavaPairRDD(riakBucket(bucketName, bucketType, rdmf).rdd, keyClass, valueClass)
  }

  def riakBucket[K <: String, V](bucketName: String, bucketType: String,
                                 keyClass: Class[K], valueClass: Class[V]): RiakJavaPairRDD[K, V] = {
    val rdmf = ReadPairValueDataMapper.factory(getClassTag(keyClass), getClassTag(valueClass))
    riakBucket(bucketName, bucketType, keyClass, valueClass, rdmf)
  }

  def riakBucket[K <: String, V](ns: Namespace, keyClass: Class[K], valueClass: Class[V], rdmf: ReadDataMapperFactory[(K, V)]): RiakJavaPairRDD[K, V] = {
    riakBucket(ns.getBucketNameAsString, ns.getBucketTypeAsString, keyClass, valueClass, rdmf)
  }

  def riakBucket[K <: String, V](ns: Namespace, keyClass: Class[K], valueClass: Class[V]): RiakJavaPairRDD[K, V] = {
    riakBucket(ns.getBucketNameAsString, ns.getBucketTypeAsString, keyClass, valueClass)
  }

  def riakTSTable[T](bucketName: String, targetClass: Class[T]): RiakTSJavaRDD[T] = {
    val classTag: ClassTag[T] = getClassTag(targetClass)
    val rdd: RiakTSRDD[T] = RiakTSRDD(sc, bucketName, ReadConf(sc.getConf))(classTag, RiakConnector(sc.getConf))
    new RiakTSJavaRDD[T](rdd, classTag)
  }

  def riakTSTable[T](bucketName: String, schema: StructType, targetClass: Class[T]): RiakTSJavaRDD[T] = {
    val classTag: ClassTag[T] = getClassTag(targetClass)
    val rdd: RiakTSRDD[T] = RiakTSRDD(sc, bucketName, ReadConf(sc.getConf), Option(schema))(classTag, RiakConnector(sc.getConf))
    new RiakTSJavaRDD[T](rdd, classTag)
  }

}

